allow ansible to connect to docker containers (without using ssh)

objectified opened the pull request:
This connection plugin allows Ansible to connect to Docker containers through using `docker exec`, to allow Ansible to configure Docker containers without first copying Ansible and your playbooks to the container (as long as there's Python, you're good to go - just like with regular servers). 

It was largely based on the work done by Lorin Hochstein on his https://github.com/lorin/ansible-docker-connection project.

Some remarks about the implementation:
- it does not use docker-py, but calls the docker CLI through subprocess; I've found that from experience with other projects that use docker-py, it's very much a moving target at the moment, and that tends to break integration (as it did quite recently in Ansible as well I believe). If people insist on using docker-py instead, I can rewrite the code accordingly.
- to copy files over to the container it `cat`s the file and writes them to the stdin of the container; this is because `docker cp` does not actually support copying from the host to the container, only the other way around. This trick/hack seems to work well enough for now, and we can gradually get rid of it once https://github.com/docker/docker/pull/13171 has been integrated into Docker.

amenonsen commented:
Note PR #7971, which uses nsenter to achieve similar functionality. There's some discussion there from @abadger about whether to use docker exec instead.

objectified commented:
@amenonsen : I see, thanks for pointing this out. As far as the discussion goes, the author of nsenter also currently seems to suggest to go with `docker exec` instead. In other regards, our PRs seem to be very similar.

abadger commented:
@lorin, @mscherer This seems to be inline with the results of the discussion that @mscherer and I had about the nsenter conneciton plugin at pycon so in that way it seems like a good idea.  Do you guys like this or have any suggestions for it?

@bcoca: I think you were updating the existing connection plugins to be compatible with devel.  Does this look okay in that regard?

I'll have a few inline comments that I'll add as well from just looking at the code.

abadger commented on lib/ansible/plugins/connections/docker.py:
```diff
@@ -0,0 +1,104 @@
+
+        self._display.vvv("FETCH %s TO %s" % (in_path, out_path), host=self._connection_info.remote_addr)
+        p = subprocess.Popen(args, stdin=subprocess.PIPE,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
Looks like defining args is missing in the fetch_file method.

objectified replied:
Ouch, thanks for catching that. That will teach me to brutally last-minute refactor something. Fixed now.

abadger commented on lib/ansible/plugins/connections/docker.py:
```diff
@@ -0,0 +1,104 @@
+        """ Transfer a file from local to container """
+        args = [self.docker_cmd, "exec", "-i", self._connection_info.remote_addr, "bash", "-c",
+                "cat > %s" % format(out_path)]
+
```
We used dd rather than cat in the jail, chroot, and zone plugins.  Can use similar code here.  I believe that code also does this without a shell which could be nice for making this more generic (not that most installs will be lacking bash but it's always nice to avoid) and potentially avoiding shell quoting issues.

mscherer commented on lib/ansible/plugins/connections/docker.py:
```diff
@@ -0,0 +1,104 @@
+        # HACK: Due to a race condition, this sometimes returns before
+        # the file has been written to disk, so we sleep for one second
+        time.sleep(1)
+
```
this doesn't sound good, since 1 second is quite a lot since we copy file in the container each time we run a module. Can you explain a bit the issue ?

objectified replied:
This was in the original code that I forked from; to be honest I'm not sure about the background of this behavior. I'll do some experiments with this as soon as I have time. @lorin could you by any chance provide some more information about this in the meantime?

lorin replied:
I don't have much to add beyond what's already written in the comments. When last I tested this code, if the sleep statement isn't there then sometimes the file will not be written to disk yet when the next task in the play executes, which can lead to unexpected playbook failures.

mscherer replied:
Wouldn't it be handled by p.wait() ? According to the documentation, this wait until the process end, so it should work.

objectified replied:
@mscherer: yes, just as execution should block by using communicate() as well (I'm assuming you meant that too)

I've been running a number of experiments with this code (without the `sleep(1)`), and I can't reproduce the described behavior, even when purposefully using larger modules (so it needs to transfer more data before continuing).

@lorin : I'm still a bit worried about this though. Could it be that this was needed because of things that you were executing (e.g. waiting for something to finish on the remote system/container), as opposed to the module not being copied entirely before it being called?

lorin replied:
@objectified I agree that communicate should block until completion, which means that the sleep shouldn't be necessary. I must admit that I didn't really dig into why I was having issues: I couldn't reproduce the problem reliably, sometimes it happened at sometimes it didn't.

If you can't reproduce the problem, then I'd suggest just removing the sleep statement, and maybe add some docs to the effect that this is an experimental plugin?

objectified replied:
@lorin Thanks for the elaboration, I'll keep an eye open for any issues that might be caused by this.

For now I've removed the sleep call. Marking the plugin as experimental might be a good idea as well, although the rest of the functionality seems to work well for me (although it hasn't been battle tested of course).

Thanks again for your attention on this.

objectified commented:
A quick wrap up in response to the comments and discussion we had:
- I've changed the implementation of put_file from `cat` to `dd`
- I've removed the `sleep` in `put_file` as I could not reproduce the race condition during a number of experiments

If there is anything else you want changed/fixed/etc., please let me know.

objectified commented:
@bcoca : any chance you could take a look at the plugin code to see if it would be compatible? If there's a reference somewhere that I should comply to instead, I'd be happy to hear it as well.

Thanks.

objectified committed:
allow ansible to connect to docker containers

objectified commented:
As discussed on IRC I've changed _connection_info to the new _play_context in the plugin. Additionally I've added the Ansible license header, and squashed the PR to one commit.

Let me know if you need anything else.

mscherer commented:
Latest docker support docker cp to fetch data out, which would be more robust that relying on dd/bash being in the container. Not sure if this can be detected somehow ( like version )

abadger commented:
Could possibly detect with something like this (naturally, caching whether we have cp or not):

```
$ docker cp --help &>/dev/null ; echo $?                                            (08:49:27)
0
[pts/27@roan /var/tmp]$ docker qw --help &>/dev/null ; echo $?                                            (08:49:31)
1
```

EDIT: Ah, I missed the fact that previous docker also supported cp... just didn't support copying data out with that command.

abadger commented on lib/ansible/plugins/connections/docker.py:
```diff
@@ -0,0 +1,119 @@
+    def _connect(self, port=None):
+        """ Connect to the container. Nothing to do """
+        return self
+
```
I would do something like local.py  does here:

``` python
        if not self._connected:
            self._display.vvv("ESTABLISH DOCKER CONNECTION FOR USER: {0}".format(self._play_context.remote_user,      host=self._play_context.remote_addr))
            self._connected = True
        return self
```

It adds logging which is a good thing to have when things go wrong.  Setting self._connected is unnecessary for docker but it is nice if someone later decides to cargo-cult from the docker module and their connection plugin does need to set the _connected attribute.

abadger commented on lib/ansible/plugins/connections/docker.py:
```diff
@@ -0,0 +1,119 @@
+                     su_user=None):
+
+        """ Run a command on the local host """
+
```
I would call super here like this:

``` python
super(Connection, self).exec_command(cmd, tmp_path, in_data=in_data, sudoable=sudoable)
```

It doesn't do anything useful for the docker connection but once again, it's nice to do it so that anyone looking at the docker connection plugin for a basis to create their own connection plugin will do the right thing (the base class's exec_command() fetch_file(), and put_file() use the @ensure_connect decorator to make sure that our connection is open before attempting these actions.  For plugins that actually need to establish a connection, this is a necessary step).

abadger commented on lib/ansible/plugins/connections/docker.py:
```diff
@@ -0,0 +1,119 @@
+    def exec_command(self, cmd, tmp_path, sudo_user=None, sudoable=False,
+                     executable='/bin/sh', in_data=None, su=None,
+                     su_user=None):
+
```
I think the signature of exec_command has changed in v2.  If you take a look at local.py or the ConnectionBase class you'll see:

``` python
    def exec_command(self, cmd, tmp_path, in_data=None, sudoable=True):
```

objectified committed:
use docker cp when docker >=1.8.0

objectified commented:
@abadger , @mscherer : I've added code that checks for the version of the docker container, and depending on that, it will use either `docker cp` or the `dd` trick to copy from local to container. I've tested this code on a number of different docker daemon versions, and it seems to work well. 

@abadger : thanks for all your input on this! I'll get it into the PR hopefully later today.

objectified committed:
fake being connected for logging purposes

objectified committed:
align exec_command() definition with local.py

objectified commented:
@abadger : both the exec_command() definition and suggested logging are implemented (and re-tested locally) now as well.

abadger commented:
@objectified Looks good!  My docker setup seems to be a little screwed up right now but to the extent I was able to test this it worked! (tested raw module with busybox.  fetch, copy, and script with a fedora image).

I'm sure there's other bugs lurking in the code -- would you be willing to work on any of those that come up?  If so, add your github id in the header somewhere (perhaps just a comment line for maintainer with your name and github id).  That way we'll know who to CC if problems are discovered.

I'll merge this soon in either case.  It seems to be working.

abadger commented:
Oops, i forgot one other hting -- the version check fails here (the distro package I have is putting some non-integer numbers into the version string.)  Here was my quick and dirty fix for it: https://gist.github.com/abadger/e9a1fa76d725ed864112

Not my best work (if there's multiple alpha versions in both version strings it could compare wrong) but I think it will work for the things that will get into the docker version string.

objectified committed:
added Maintainer comment header

objectified committed:
replace compare_versions() with distutils.version

objectified commented:
@abadger : That's great news, thanks! And yes, I'll definitely want to keep working on this.

As for the requests: 
- I've added a Maintainer header
- I've replaced the compare_versions() function with an implementation that uses distutils.version.LooseVersion; I ran into it recently while doing some work, and I noticed that it was also the way that is often used in Ansible modules. This implementation is likely to be much more robust (and less code in the plugin). I've tested with a few different Docker versions, and it seems to work well. If you could test this against your setup, that would be great.

abadger commented:
Thanks @objectified !  

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
